src/Fhir.Anonymizer.Shared.Core/Utility/SubstituteUtility.cs

```diff
@@ -0,0 +1,126 @@
+                    else
+                    {
+                        // Remove source node when no target node availabe and we don't need to keep the source node
+                        node.Remove(child);
+                        processResult.AddProcessRecord(AnonymizationOperations.Substitute, child);
+                    }
+                }
+
+                while (i < targetChildren.Count)
+                {
+                    // Add extra target nodes, create a new copy before adding
+                    node.Add(s_provider, ElementNode.FromElement(targetChildren[i ++]));
+                }
+            }
+
+            // children nodes not presented in replacement value, we need either remove or keep a dummy copy
+            var nonReplacementChildren = node.Children()
+                .Where(element => !replaceChildrenNames.Contains(element.Name))
+                .Cast<ElementNode>().ToList();
+            foreach (var child in nonReplacementChildren)
+            {
+                if (visitedNodes.Contains(child))
+                {
+                    continue;
+                }
+                else if (keepNodes.Contains(child))
+                {
+                    SubstituteNode(child, GetDummyNode(), visitedNodes, keepNodes);
+                }
+                else
+                {
+                    node.Remove(child);
+                }
+            }
+
+            node.Value = replacementNode.Value;
+            processResult.AddProcessRecord(AnonymizationOperations.Substitute, node);
+            return processResult;
+        }
+
+        // To keep consistent anonymization changes made by preceding rules, we should figure out whether a node can be removed during substitution
+        public static bool ShouldKeepNodeDuringSubstitution(ElementNode node, HashSet<ElementNode> visitedNodes, HashSet<ElementNode> keepNodes)
+        {
+            var shouldKeep = false;
+            // If a child (no matter how deep) has been modified, this node should be kept
+            foreach (var child in node.Children().Cast<ElementNode>())
+            {
+                shouldKeep |= ShouldKeepNodeDuringSubstitution(child, visitedNodes, keepNodes);
+            }
+
+            // If this node its self has been modified, it should be kept
+            if (shouldKeep || visitedNodes.Contains(node))
+            {
+                keepNodes.Add(node);
+                return true;
+            }
+
+            return shouldKeep;
+        }
+
+        // Post-process to mark all substituted children nodes as visited
+        public static void MarkSubstitutedFragementAsVisited(ElementNode node, HashSet<ElementNode> visitedNodes)
+        {
+            visitedNodes.Add(node);
+            foreach (var child in node.Children().Cast<ElementNode>())
+            {
+                MarkSubstitutedFragementAsVisited(child, visitedNodes);
+            }
+        }
+
+        private static ElementNode GetDummyNode()
```

Comment:
If we generate the replace node with empty field, should we still need dummy node here?

Reply:
Actually we can replace the node value to null and remove all children that are not either vistitedNode or keepNode. The result will be the same to use a dummy node. I prefer to use a dummy node as it reuses the substitution logic